### PR TITLE
OCSP multi

### PR DESCRIPTION
**Pull request by vdukhovni**
Implement multi-process OCSP responder.
    
With "-multi" the OCSP responder forks multiple child processes, and respawns them as needed.  This can be used as a long-running service, not just a demo program.  Therefore the index file is automatically re-read when changed.  The responder also now optionally times out client requests.

### CODE REVIEW

File: apps/ocsp.c [hunk 1]
```diff
@@ -746,6 +806,175 @@ int ocsp_main(int argc, char **argv)
...
+    return ilen;
+}
+
+static int index_changed(CA_DB *rdb)
+{
+    struct stat sb;
+
+    if (rdb && stat(rdb->dbfname, &sb) != -1) {
+        if (rdb->dbst.st_mtime != sb.st_mtime ||
+            rdb->dbst.st_ctime != sb.st_ctime ||
```
Comment (mattcaswell) [inline apps/ocsp.c]: Style: Put the `||` operators at the beginning of the next line, not the end of the previous one.

File: apps/ocsp.c [hunk 2]
```diff
@@ -746,6 +806,175 @@ int ocsp_main(int argc, char **argv)
...
+}
+
+static void killall(int ret, pid_t *kidpids)
+{
+    int i;
+
+    for (i = 0; i < multi; ++i)
+        if (kidpids[i] != 0)
+            (void) kill(kidpids[i], SIGTERM);
+    sleep(1);
```
Comment (mattcaswell) [inline apps/ocsp.c]: Why this?

File: apps/ocsp.c [hunk 3]
```diff
@@ -746,6 +806,175 @@ int ocsp_main(int argc, char **argv)
...
+                }
+                if (status != 0) {
+                    if (WIFEXITED(status))
+                        syslog(LOG_WARNING, "child process: %ld, exit status: %d",
+                               (long)fpid, WEXITSTATUS(status));
+                    else if (WIFSIGNALED(status))
+                        syslog(LOG_WARNING, "child process: %ld, term signal %d%s",
+                               (long)fpid, WTERMSIG(status),
+                               WCOREDUMP(status) ? " (core dumped)" : "");
+                    sleep(1);
```
Comment (mattcaswell) [inline apps/ocsp.c]: Why this?

File: apps/ocsp.c [hunk 4]
```diff
@@ -746,6 +806,175 @@ int ocsp_main(int argc, char **argv)
...
+                syslog(LOG_ERR, "fatal: waitpid(): %m");
+                killall(1, kidpids);
+            }
+        }
+        if (termsig)
+            break;
+
+        switch(fpid = fork()) {
+        case -1:            /* error */
+            sleep(30);
```
Comment (mattcaswell) [inline apps/ocsp.c]: Please can you explain this?

File: apps/ocsp.c [hunk 5]
```diff
@@ -746,6 +806,175 @@ int ocsp_main(int argc, char **argv)
...
+            break;
+        case 0:             /* child */
+            RAND_poll();
```
Comment (mattcaswell) [inline apps/ocsp.c]: Perhaps we should fail if RAND_poll() fails?

File: apps/ocsp.c [hunk 6]
```diff
@@ -1083,7 +1310,16 @@ static int urldecode(char *p)
 }
 # endif
 
-static int do_responder(OCSP_REQUEST **preq, BIO **pcbio, BIO *acbio)
+# ifdef OCSP_DAEMON
+static void sock_timeout(int signum)
+{
+    if (acfd != (int) INVALID_SOCKET)
```
Comment (mattcaswell) [inline apps/ocsp.c]: Style nit: No space before `INVALID_SOCKET`

File: apps/ocsp.c [hunk 7]
```diff
@@ -1083,7 +1310,16 @@ static int urldecode(char *p)
...
+        (void) shutdown(acfd, SHUT_RD);
```
Comment (mattcaswell) [inline apps/ocsp.c]: Style nit: No space before shutdown.

File: apps/ocsp.c [hunk 8]
```diff
@@ -1161,13 +1421,17 @@ static int do_responder(OCSP_REQUEST **preq, BIO **pcbio, BIO *acbio)
...
+    if (req == NULL)
+        log_message(LOG_ERR, "Error parsing OCSP request");
 
     *preq = req;
 
+out:
+#  ifdef OCSP_DAEMON
+    if (timeout > 0)
+        alarm(0);
+    acfd = (int) INVALID_SOCKET;
```
Comment (mattcaswell) [inline apps/ocsp.c]: No space before `INVALID_SOCKET`

**Comment by richsalz** [comment]
I seem to recall someone pushing back against large undiscussed changes, if only I could remember who it was... :)

There are issues with openssl and fork, #4852 etc.  Are we sure this code is safe?

**Comment by mattcaswell** [comment]
> There are issues with openssl and fork, #4852 etc. 

That particular issue is to do with fork + multi-thread IIUC - which is not the case here.

File: apps/apps.h [hunk 9]
```diff
@@ -14,6 +14,14 @@
 # include "internal/nelem.h"
 # include <assert.h>
 
+# ifndef NO_SYS_TYPES_H
```
[the remark this answers is not in the document]
Comment (vdukhovni) [reply apps/apps.h]: Thanks. Removed.

File: apps/ocsp.c [hunk 10]
```diff
@@ -34,6 +35,22 @@ NON_EMPTY_TRANSLATION_UNIT
 # include <openssl/bn.h>
 # include <openssl/x509v3.h>
 
+# if (defined(OPENSSL_SYS_UNIX) || defined(OPENSSL_SYS_LINUX)) && !defined(OPENSSL_NO_SOCK)
```
[the remark this answers is not in the document]
Comment (vdukhovni) [reply apps/ocsp.c]: Thanks, simplified.

File: apps/ocsp.c [hunk 11]
```diff
@@ -81,7 +109,10 @@ typedef enum OPTION_choice {
     OPT_INDEX, OPT_CA, OPT_NMIN, OPT_REQUEST, OPT_NDAYS, OPT_RSIGNER,
     OPT_RKEY, OPT_ROTHER, OPT_RMD, OPT_RSIGOPT, OPT_HEADER,
     OPT_V_ENUM,
-    OPT_MD
+    OPT_MD,
+# ifdef OCSP_DAEMON
```
[the remark this answers is not in the document]
Comment (vdukhovni) [reply apps/ocsp.c]: Thanks. Removed.

File: apps/ocsp.c [hunk 12]
```diff
@@ -514,14 +551,52 @@ int ocsp_main(int argc, char **argv)
         if (rkey == NULL)
             goto end;
     }
+
+    if (ridx_filename && (!rkey || !rsigner || !rca_cert)) {
```
[the remark this answers is not in the document]
Comment (vdukhovni) [reply apps/ocsp.c]: Thanks. `!= NULL` added.

File: apps/ocsp.c (hunk 1 above)
Comment (vdukhovni) [reply apps/ocsp.c]: Done.

File: apps/ocsp.c (hunk 4 above)
Comment (vdukhovni) [reply apps/ocsp.c]: When `fork()` fails, the system is severely resource starved.  It is best to back off and wait before trying again.

File: apps/ocsp.c (hunk 5 above)
Comment (vdukhovni) [reply apps/ocsp.c]: I don't think this is critical for OCSP, though if we're signing with ECDSA perhaps not so good...  What do you think?

File: apps/ocsp.c (hunk 6 above)
Comment (vdukhovni) [reply apps/ocsp.c]: Done.

File: apps/ocsp.c (hunk 7 above)
Comment (vdukhovni) [reply apps/ocsp.c]: Done.

File: apps/ocsp.c (hunk 8 above)
Comment (vdukhovni) [reply apps/ocsp.c]: Done.

**Comment by vdukhovni** [comment]
Pushed fixup commit.

**Comment by vdukhovni** [comment]
Pushed fixup commit for warnings and Windows build error

File: apps/ocsp.c [hunk 13]
```diff
@@ -514,14 +550,53 @@ int ocsp_main(int argc, char **argv)
...
+    if (ridx_filename != NULL
+        && (rkey != NULL || rsigner != NULL || rca_cert != NULL)) {
+        BIO_printf(bio_err,
+                   "Responder mode requires certificate, key, and CA.\n");
+        goto end;
+    }
+
+    if (ridx_filename != NULL) {
+        rdb = load_index(ridx_filename, NULL);
+        if (!rdb || !index_index(rdb)) {
```
Comment (mattcaswell) [inline apps/ocsp.c]: Nit: `rdb == NULL`

File: apps/ocsp.c [hunk 14]
```diff
@@ -746,6 +806,177 @@ int ocsp_main(int argc, char **argv)
...
+
+    va_start(ap, fmt);
+# ifdef OCSP_DAEMON
+    if (multi) {
+        vsyslog(level, fmt, ap);
+        if (level >= LOG_ERR)
+            ERR_print_errors_cb(print_syslog, &level);
+    }
+# endif
+    if (! multi) {
```
Comment (mattcaswell) [inline apps/ocsp.c]: Nit: No space between `!` and `multi`

File: apps/ocsp.c [hunk 15]
```diff
@@ -746,6 +806,177 @@ int ocsp_main(int argc, char **argv)
...
+    syslog(level, "%.*s", ilen, str);
+
+    return ilen;
+}
+
+static int index_changed(CA_DB *rdb)
+{
+    struct stat sb;
+
+    if (rdb && stat(rdb->dbfname, &sb) != -1) {
```
Comment (mattcaswell) [inline apps/ocsp.c]: Nit: `rdb != NULL`

File: apps/ocsp.c [hunk 16]
```diff
@@ -746,6 +806,177 @@ int ocsp_main(int argc, char **argv)
...
+    return 0;
+}
+
+static void killall(int ret, pid_t *kidpids)
+{
+    int i;
+
+    for (i = 0; i < multi; ++i)
+        if (kidpids[i] != 0)
+            (void) kill(kidpids[i], SIGTERM);
```
Comment (mattcaswell) [inline apps/ocsp.c]: Nit: no space in cast

**Comment by vdukhovni** [comment]
> @richsalz I seem to recall someone pushing back against large undiscussed changes, if only I could remember who it was... :)

Yes, that was me.  I don't see the "ocsp -multi" feature as a large change.  It is is not in library code, and is rather simple, just prefork and respawn some children to handle OCSP requests, reload the index if needed, and support a timeout.  The quantity of code involved is rather small and straight-forward.

Yes, this is somewhat late, sorry about that. A bit too swamped to get it out earlier, and the deadlined helped to move it to the top of the stack. :-(

File: apps/ocsp.c (hunk 5 above)
Comment (mattcaswell) [reply apps/ocsp.c]: I think it is worthwhile. It is probably good practice for us to do this even if we can't currently determine whether it is critical or not. Who knows what changes we might add in the future which might suddenly make it critical?

File: apps/ocsp.c (hunk 13 above)
Comment (vdukhovni) [reply apps/ocsp.c]: Missed that one. Thanks.

File: apps/ocsp.c (hunk 14 above)
Comment (vdukhovni) [reply apps/ocsp.c]: Done.

File: apps/ocsp.c (hunk 15 above)
Comment (vdukhovni) [reply apps/ocsp.c]: Done.

File: apps/ocsp.c (hunk 16 above)
Comment (vdukhovni) [reply apps/ocsp.c]: Sadly, I am used to spaces after casts from Postfix, will have to remember not to for OpenSSL...  Fixed.

**Comment by vdukhovni** [comment]
Thanks for the continuing review. Pushed commit fixing style nits, adding CHANGES entry, and testing for RAND_poll() success.

**Comment by vdukhovni** [comment]
I'd like to squash everything down to one commit at the end, let me know whether you'd like to review that again before the merge...

**Comment by vdukhovni** [comment]
Thanks Matt!  Much appreciated.  Pushed and closed.

Thanks RIch, to be honest the tone was a bit difficult to discern, so the clarification is welcome!

**Pull request closed without merging**